clr-installer/main.go

```diff
@@ -123,57 +126,70 @@ func main() {
 
 		os.Exit(1)
 	}
-	_ = f.Close()
-}
 
-// execute is called by main to begin execution of the installer
-//nolint: gocyclo  // TODO: Refactor this
-func execute(options args.Args) error {
-	var err error
-
-	if options.DemoMode {
-		model.Version = model.DemoVersion
-	}
-	// Make the Version of the program visible to telemetry
-	telemetry.ProgVersion = model.Version
+}
 
-	log.Info(path.Base(os.Args[0]) + ": " + model.Version +
-		", built on " + model.BuildDate)
+func callFrontEnd(options args.Args, md *model.SystemInstall, installReboot *bool,
+	rootDir string, errChan chan error, done chan bool) {
 
-	if options.PamSalt != "" {
-		if status, err := user.IsValidPassword(options.PamSalt); !status {
-			return fmt.Errorf(err)
+	var err error
+	for _, fe := range frontEndImpls {
+		if !fe.MustRun(&options) {
+			continue
 		}
 
-		hashed, errHash := encrypt.Crypt(options.PamSalt)
-		if errHash != nil {
-			return errHash
+		*installReboot, err = fe.Run(md, rootDir, options)
+		if err != nil {
+			feName := classExp.FindString(reflect.TypeOf(fe).String())
+			if feName == "" {
+				feName = "unknown"
+			}
+			if errLog := md.Telemetry.LogRecord(feName, 3, err.Error()); errLog != nil {
+				log.Error("Failed to log Telemetry fail record: %s", feName)
+			}
+
+			if errors.IsValidationError(err) {
+				fmt.Println("Error: Invalid configuration:")
+				errChan <- err
+			} else {
+				log.RequestCrashInfo()
+				errChan <- err
+			}
 		}
 
-		fmt.Println(hashed)
-		return nil
+		break
 	}
 
-	if options.Version {
-		fmt.Println(path.Base(os.Args[0]) + ": " + model.Version)
-		return nil
+	done <- true
+
+}
+
+func callTelemetryTerminate(md *model.SystemInstall, done chan bool, sigs chan os.Signal) {
```

Comment:
This would be better named 
```suggestion
func handleSignals(md *model.SystemInstall, done chan bool, sigs chan os.Signal) {
```

Reply:
Sounds good. Done